### DISCUSSION TITLE
BIO* pub = BIO_new(BIO_s_mem()); is failing in openssl 3.0 it worked fine with openssl 1.0.2zg

### DISCUSSION THREAD

**Discussion by aniljadaun** [openssl/openssl#22413] [Q&A]
Hi 

BIO* pub = BIO_new(BIO_s_mem()); is not allocating any memory and ERR_print_errors_fp doesn't log any error . Can you please help me in getting this issue diagnosed ? i am using openssl 3.1 with inbuilt legacy module.

**Comment by mattcaswell** [comment]
So, by "is not allocating any memory", do you mean that the `BIO_new` call returns NULL? There aren't very many reasons that would occur without logging an error and none of them seem particularly likely!!

**Reply by aniljadaun** [reply]
yes that is right ,BIO_new returns null. Not sure how to debug this any way you suggest to  diagnose the issue ?

**Reply by mattcaswell** [reply]
This is quite surprising. The `BIO_new` function is just a simple wrapper around the `BIO_new_ex` function which is defined here:

https://github.com/openssl/openssl/blob/f4bc5568a973e640a70f416da37a030c25a38ebc/crypto/bio/bio_lib.c#L81-L124

As you can see most failure reasons do raise an error on failure (which should be visible via `ERR_print_errors_fp`). The one that doesn't is if the call to `CRYPTO_new_ex_data` fails. But taking the analysis further and looking at reasons why that might fail without an error only shows errors which we wouldn't expect to happen (e.g. malloc failure etc).

So, I'm at a bit of loss as to why you are seeing this, but it points to some fundamental problem.

Where did you get your copy of OpenSSL from? Did you download/build it yourself, or did it come from your OS or other 3rd party? The only way I can really suggest to further investigate this is to build a debug copy of OpenSSL and step through the BIO_new_ex function in the debugger and see where it fails.

**Comment by nhorman** [comment]
@aniljadaun additionally, can you share some reproducer code for this?
